packages/flutter/lib/src/widgets/async.dart

```diff
@@ -0,0 +1,387 @@
+        });
+      }, onDone: () {
+        setState(() {
+          _summary = config.afterDone(_summary);
+        });
+      });
+      _summary = config.afterConnected(_summary);
+    }
+  }
+
+  void _unsubscribe() {
+    if (_subscription != null) {
+      _subscription.cancel();
+      _subscription = null;
+    }
+  }
+}
+
+/// The state of connection to an asynchronous computation.
+///
+/// See also:
+///
+/// * [AsyncSnapshot]
+enum ConnectionState {
+  /// Not currently connected to any asynchronous computation.
+  none,
+
+  /// Connected to an asynchronous computation and awaiting interaction.
+  waiting,
+
+  /// Connected to an active asynchronous computation.
+  active,
+
+  /// Connected to a terminated asynchronous computation.
+  done
+}
+
+/// Snapshot of an asynchronous interaction.
+///
+/// Used by widget builders that depend only on the most recent interaction with
+/// an asynchronous computation.
+///
+/// See also:
+///
+/// * [StreamBuilder]
+/// * [FutureBuilder]
+class AsyncSnapshot<T> {
+  /// Creates an [AsyncSnapshot] with the specified [connectionState],
+  /// and optionally either [data] or [error] (but not both).
+  AsyncSnapshot(this.connectionState, this.data, this.error) {
+    assert(connectionState != null);
+    assert(data == null || error == null);
+  }
+
+  /// Creates an [AsyncSnapshot] in [ConnectionState.none] with `null` data and
+  /// `error`.
+  factory AsyncSnapshot.nothing() => new AsyncSnapshot<T>(ConnectionState.none, null, null);
+
+  /// Creates an [AsyncSnapshot] in [ConnectionState.active] with the specified
+  /// [data].
+  factory AsyncSnapshot.activeData(T data) => new AsyncSnapshot<T>(ConnectionState.active, data, null);
+
+  /// Creates an [AsyncSnapshot] in [ConnectionState.active] with the specified
+  /// [error].
+  factory AsyncSnapshot.activeError(Object error) => new AsyncSnapshot<T>(ConnectionState.active, null, error);
+
+  /// Current state of connection to the asynchronous computation.
+  final ConnectionState connectionState;
+
+  /// Latest data received. Is null, if [error] is not.
+  final T data;
+
+  /// Latest error object received. Is null, if [data] is not.
+  final Object error;
+
+  /// Returns a snapshot like this one, but in the specified [state].
+  AsyncSnapshot<T> inState(ConnectionState state) => new AsyncSnapshot<T>(state, data, error);
+
+  /// Returns whether this snapshot contains a non-null data value.
+  bool get hasData => data != null;
+
+  /// Returns whether this snapshot contains a non-null error value.
+  bool get hasError => error != null;
+
+  @override
+  String toString() => 'AsyncSnapshot($connectionState, $data, $error)';
+
+  @override
+  bool operator ==(dynamic other) {
+    if (identical(this, other))
+      return true;
+    if (other is! AsyncSnapshot<T>)
+      return false;
+    final AsyncSnapshot<T> typedOther = other;
+    return connectionState == typedOther.connectionState &&
+      data == typedOther.data &&
```

Comment:
really trivial nit: please indent the subparts of this expression so that they line up (7 indent instead of 2, basically).

Or you can be really fancy as we do in some places and have the `&&` operators lined up with the `rn` of `return`. :-)